Switchdev Network Configure

wmousa opened the pull request:
This repo describes the sriov and network configuration that needs
to occur on any node that need to configure sriov_pfs with switchdev
capability and vf-lag with bonding

wmousa committed:
Switchdev Network Configure
This repo describes the sriov and network configuration that needs
to occur on any node that need to configure sriov_pfs with switchdev
capability and vf-lag with bonding

moshe010 commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+#!/usr/bin/python3
```
why python3 ? centos 7 comes with python 2.7

adrianchiris commented on switchdev_utils/python/README.md:
```diff
@@ -0,0 +1,41 @@
+## Switchdev Network Configure  
```
Switchdev Network Configuration Service

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/README.md:
```diff
@@ -0,0 +1,41 @@
+## Switchdev Network Configure  
+This repo describes the sriov and network configuration that needs to occur on any node that need to configure sriov pfs with switchdev capability and vf-lag with bonding  
```
Suggested reword

> This repo describes the sriov and network configuration that needs to occur on any node that need to configure sriov pfs with switchdev capability and vf-lag with bonding

This folder contains SR-IOV, OVS and network configuration service for switchdev and VF-LAG

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/README.md:
```diff
@@ -0,0 +1,41 @@
+## Switchdev Network Configure  
+This repo describes the sriov and network configuration that needs to occur on any node that need to configure sriov pfs with switchdev capability and vf-lag with bonding  
+
+The configuration listed below shall be performed as part of a systemd script triggered during machine boot and IFCFG files for network confiuration  
```
confiuration -> configuration

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/README.md:
```diff
@@ -0,0 +1,41 @@
+- Create OVS bridges  
+- Enable hw-offload in OVS  
+- Connect bond to bridge  
+
```
add : Each configuration step is optionally performed on specified interfaces via configuration file.

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/README.md:
```diff
@@ -0,0 +1,41 @@
+List of ovs bridges which need to be created and configured  
+    - name: The name of ovs bridge  
+    - ports: List of ovs bridge ports (sriov pfs or linux bind)  
+
```
you should add a system prerequisite section as well
as you assume that NIC Firmware SR-IOV parameters are already configured and that Mstflint package is installed

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+#!/usr/bin/python
```
@wmousa , was this validated to work with both python 2 and 3 ?

wmousa replied:
yes

moshe010 replied:
yes

adrianchiris commented on switchdev_utils/python/README.md:
```diff
@@ -0,0 +1,41 @@
+- ovsBridges:  
+List of ovs bridges which need to be created and configured  
+    - name: The name of ovs bridge  
+    - ports: List of ovs bridge ports (sriov pfs or linux bind)  
```
> or linux bind
or linux bonds

>ports: List of ovs bridge ports (sriov pfs or linux bind)

so ports should be a subset of sriov pfs or linux bonds. i dont think the code enforces that right ? we should imo.

wmousa replied:
Done,
Currently no, but will think about it when do the validation code

adrianchiris commented on switchdev_utils/python/README.md:
```diff
@@ -0,0 +1,41 @@
+List of SRIOV PF objects which need to be configured, and has the following attributes:  
+    - name: The name of the sriov pf interface  
+    - numOfVfs: Number of vfs to be confgiured on the sriov pf  
+    - switchdev: Boolean value where to enable switchdev or disable it  
```
setting switchdev to false will not actively disable switchdev right ? so i would:

- switchdev: boolean, Enable switchdev mode.

wmousa replied:
Yes, done

moshe010 replied:
I don't think we need it. anyway numofVFs =0 can disable it

adrianchiris commented on switchdev_utils/python/README.md:
```diff
@@ -0,0 +1,41 @@
+    - name: The name of the sriov pf interface  
+    - numOfVfs: Number of vfs to be confgiured on the sriov pf  
+    - switchdev: Boolean value where to enable switchdev or disable it  
+    - vfTrust: Boolean value where to configure vf trust or not  
```
same comment as above as we do not actively disable vf trust

wmousa replied:
done

moshe010 replied:
Because it for now just to mavenir we can remove this option and to implicitly in the code. for the dicussion we had today it is not the regular VF trust

adrianchiris commented on switchdev_utils/python/README.md:
```diff
@@ -0,0 +1,41 @@
+- linuxBonds:  
+list of linux bond objects which need to be created and configured  
+    - name: The name of the bond interface  
+    - slaves: List of bond slaves interfaces (sriov pfs)  
```
> slaves: List of bond slaves interfaces (sriov pfs)  

so slaves should be a subset of sriov pfs i dont think the code enforces that right ? we should imo.

wmousa replied:
Currently no, but will think about it when do the validation code

adrianchiris commented on switchdev_utils/python/README.md:
````diff
@@ -0,0 +1,41 @@
+- Run the script  
+```$ /etc/switchdev_network_config```  
+
+
````
nit: too many new lines

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/README.md:
````diff
@@ -0,0 +1,41 @@
+```$ vi /etc/config.yaml```  
+- Run the script  
+```$ /etc/switchdev_network_config```  
+
````
can you add a note that when executed it will create the following systemd services `/etc/systemd/system/sriov_config.service`
`/etc/systemd/system/sriov_bind.service`

adrianchiris replied:
just making sure, no reboot is needed after right ?

wmousa replied:
yes

adrianchiris commented on switchdev_utils/python/README.md:
```diff
@@ -0,0 +1,41 @@
+List of ovs bridges which need to be created and configured  
+    - name: The name of ovs bridge  
+    - ports: List of ovs bridge ports (sriov pfs or linux bind)  
+
```
can you add an example in the README ? with a textual explanation what it configures, to help clarify the config file and expected results upon reboot

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+MAX_RETRIES = 10
+
+SRIOV_CONFIG_SERVICE_FILE = "/etc/systemd/system/sriov_config.service"
+SRIOV_CONFIG_DEVICE_CONTENT = """[Unit]
```
`SRIOV_CONFIG_SERVICE_CONTENT` ?

wmousa replied:
Done

adrianchiris commented on switchdev_utils/python/README.md:
````diff
@@ -0,0 +1,41 @@
+- Run the script  
+```$ /etc/switchdev_network_config```  
+
+
````
perhaps a section on how to Uninstall?

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+SRIOV_BIND_SERVICE_FILE = "/etc/systemd/system/sriov_bind.service"
+SRIOV_BIND_SERVICE_CONTENT = """[Unit]
+Description=SR-IOV vfs binding
+After=network.service sriov_config.service
```
no need for `Before=openvswitch.service` ?

wmousa replied:
no need

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+After=network.service sriov_config.service
+[Service]
+Type=oneshot
+ExecStart=/etc/switchdev_network_config.py sriov_bind
```
filename is without `.py`

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+"""
+
+all_nics_map = OrderedDict()
+mlnx_vfs_pcis_list = sriov_pfs = bonds = bridges = []
```
at L#54 they are all pointing to the same list.
appending to one will reflect in all others, can you initialize independently ?

wmousa replied:
ooh yes

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+logger = logging.getLogger(__name__)
+
+
+if os.path.exists(CONFIG_FILE):
```
can you wrap this in a function and call it in main?

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+def configure_vfs_trust_switchdev(sriov_pf):
+    pf_pci = get_pf_pci(sriov_pf)
+    try:
+        cmd = ['/usr/bin/mlxreg', '-d', pf_pci, '--reg_id', '0xc007',
```
perhaps use mstreg from mstflint package instead WDYT ?

wmousa replied:
@adrianchiris  there is no mstreg :)

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+                       % sriov_pf)
+
+
+def _wait_for_uplink_rep_creation(sriov_pf):
```
no need for _ in the function name imo

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+    return pf_device_id
+
+
+def configure_vfs_trust_switchdev(sriov_pf):
```
can you add a comment here:

```
configure vHCA trust for all VFs
```

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+        cmd = ['/usr/bin/mlxreg', '-d', pf_pci, '--reg_id', '0xc007',
+               '--reg_len', '0x40', '--indexes', '0x0.0:32=0x80000000',
+               '--yes', '--set', '0x4.0:32=0x1']
+        subprocess.check_call(cmd)
```
is there a way to print the command output and error in case of RC != 0 perhaps you should wrap it as we do in first boot script (just with subprocess module and no oslo processutils):

you can use subprocess.PIPE for stdout and stderr as it not expected for the output/error to be long.


```
def run_command(*cmd):
    try:
        p = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        out, err = p.communicate()
        ret = p.returncode
        if ret != 0:
            LOG.error(.....)
            raise ...
       if err:
           LOG.warning("Got stderr output: %s" % err)
        LOG.debug(out)
        return out
    except Exception as e:
        LOG.error("Failed to execute %s, %s", ' '.join(cmd), str(e))
        raise e
```

adrianchiris replied:
then if error occurred we can have greater visibility into what happened

wmousa replied:
you made it easy 👍

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+def configure_switchdev(sriov_pf):
+    pf_pci = get_pf_pci(sriov_pf)
+    pf_device_id = get_pf_device_id(sriov_pf)
+    if pf_device_id == "0x1013" or pf_device_id == "0x1015":
```
These are ConnectX-4 and ConnectX-4Lx, can you move 0x1013 and 0x1015 to consts ?

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+
+def get_pf_pci(sriov_pf):
+    pf_pci_path = os.path.join(SYS_CLASS_NET, sriov_pf, "device/uevent")
+    pf_info = get_file_data(pf_pci_path)
```
this can be replaced with:

`return os.path.basename(os.readlink(pf_pci_path))`

then no need to read the file and use regex here

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+        with open(sriov_numvfs_path, 'r') as f:
+            return int(f.read())
+    except IOError:
+        logger.error("Unable to read numvfs for %s" % sriov_pf)
```
can you also log the error ?

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+        with open(filename, 'r') as f:
+            return f.read()
+    except IOError:
+        logger.error("Error reading file: %s" % filename)
```
can you also log the error ?

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+
+def get_file_data(filename):
+    if not os.path.exists(filename):
+        return ''
```
a debug log that the file does not exist would be nice here if you like :)

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+
+def get_vendor_id(ifname):
+    try:
+        with open(os.path.join(SYS_CLASS_NET, ifname, "device/vendor"),
```
nit: use `get_file_data()` and no need to strip the content of this file

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+            vf_info = get_file_data(os.path.join(SYS_CLASS_NET, pf_name,
+                                    "device", pf_file, "uevent"))
+            vf_pcis_list.append(re.search(r'PCI_SLOT_NAME=(.*)',
+                                          vf_info, re.MULTILINE).group(1))
```
`os.path.basename(os.readlink(....))` trick will work here as well if you like

wmousa replied:
done :)

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+        raise
+
+
+def bring_interfaces_up():
```
can you provide the net interface list as parameter ?

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+                write_config(ifcfg_config_path(port), ovs_port_data)
+
+
+def bind_vfs(sriov_bind_pcis_list=None):
```
can you add a log message at the begining of this process ?

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+SRIOV_CONFIG_DEVICE_CONTENT = """[Unit]
+Description=SR-IOV numvfs configuration
+After=systemd-udev-settle.service
+Before=openvswitch.service
```
should this run before network.service?

wmousa replied:
In general ovs has, but you are right if ovs is customized , so it may not have it in ovs service file
I'll add it

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+                                               sriov_pf.get("name"))
+            write_config(ifcfg_config_path(sriov_pf.get("name")),
+                         sriov_pf_data)
+            from_sriov_service
```
typo ?

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+        yaml.safe_dump(data, f, default_flow_style=False)
+
+
+def get_sriov_bind_pcis_list():
```
nit: `get_sriov_bind_pci_list()`
list already hints there are several.

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+
+def get_sriov_bind_pcis_list():
+    contents = get_file_data(SRIOV_BIND_CONFIG_FILE)
+    sriov_bind_pcis_list = yaml.safe_load(contents) if contents else {}
```
nit: `sriov_bind_pci_list`

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+    return sriov_bind_pcis_list
+
+
+def update_sriov_bind_pcis_list(sriov_bind_pcis_list):
```
same as above for both method name and arg

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+def configure_sriov_pfs(sriov_pfs=[], from_sriov_service=False):
+    global all_nics_map
+    global mlnx_vfs_pcis_list
+    if not sriov_pfs:
```
i see sriov_pfs is always initialized from config file, why do you need this here ?
if its not needed then `get_sriov_map_list()` is not needed as well.

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+            raise
+        # Configure switchdev mode
+        vendor_id = get_vendor_id(sriov_pf.get("name"))
+        if (sriov_pf.get("switchdev") and
```
can you log a warning in an `else` block in case someone provided switchdev configuration for a non Mellanox NIC

wmousa replied:
The else will be either none Mellanox NIC or none switchdev or both, Do you still nee to log it ? @adrianchiris

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+                         sriov_pf_data)
+            from_sriov_service
+            all_nics_map[sriov_pf.get("name")] = sriov_pf_data
+    update_sriov_bind_pcis_list(mlnx_vfs_pcis_list)
```
should you put this also under `if not from_sriov_service` ?

wmousa replied:
@adrianchiris  what if he changed the number of vfs in config file and didn't run the script again and rebooted the machine ?

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+WantedBy=multi-user.target
+"""
+
+all_nics_map = OrderedDict()
```
can you add a comment explaining what this map contain its keys and data

as i see its used for storing ifcfg file data per interface, however the name is a bit general

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+            if curr_numvfs == sriov_pf.get("numOfVfs"):
+                logger.info("Numvfs already configured for %s"
+                            % sriov_pf.get("name"))
+            elif curr_numvfs != "0":
```
curr_numvfs is int , and here you are comparing to string

wmousa replied:
sriov_pf.get("numOfVfs") should be an int, but to be sure casting it

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+    global all_nics_map
+    for linux_bond in linux_bonds:
+        bond = linux_bond.get("name")
+        bond_slaves = linux_bond.get("slaves")
```
do we want to ensure slaves are part of PFs ?

wmousa replied:
will do it in validation stage

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+        # implement ifcfg file for ovs bridge
+        bridge = ovs_bridge.get("name")
+        ovs_ports = ovs_bridge.get("ports")
+        bridge_data = "%(data)s DEVICE=%(name)s\nDEVICETYPE=ovs\n" \
```
taken from os-net-config ?

wmousa replied:
yes

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+    for ovs_bridge in ovs_bridges:
+        # implement ifcfg file for ovs bridge
+        bridge = ovs_bridge.get("name")
+        ovs_ports = ovs_bridge.get("ports")
```
should we enforce that ports are taken from either Pfs or Bonds  configuration?

wmousa replied:
will do it in validation stage

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+    subprocess.check_call(['systemctl', 'enable', 'sriov_bind'])
+
+
+def enable_hw_offloaded(hw_offload_enabled=False):
```
can you add a log message that we are enabling hw-offload for OpenVswitch ?

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+                logger.error("Failed to bind vf %s" % vf_pci)
+
+
+def configure_sriov_bind_service():
```
can you add a log message ?

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+    update_sriov_bind_pcis_list(mlnx_vfs_pcis_list)
+
+
+def configure_sriov_config_service():
```
can you add a log message ?

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+    subprocess.check_call(['systemctl', 'enable', 'sriov_config'])
+
+
+def configure_linux_bonds(linux_bonds):
```
can you add a log message ?

wmousa replied:
done

adrianchiris commented on switchdev_utils/python/switchdev_network_config:
```diff
@@ -0,0 +1,461 @@
+                write_config(ifcfg_config_path(slave), slave_data)
+
+
+def configure_ovs_bridges(ovs_bridges):
```
can you add a log message ?

wmousa replied:
done

adrianchiris reviewed (commented):
could you change the folder name to : `config-service` or `service` as it better reflects what you are doing in this folder

adrianchiris commented on switchdev_utils/config-service/README.md:
```diff
@@ -0,0 +1,55 @@
+## Prerequisites
+The below is assumed to be installed/configured on the host
+- NICs Firmware up to date  (as bundled with MLNX_OFED 5.2)
+- (MLNX_OFED 5.2 or above ) or (mft 4.16 or above)
```
we discussed using mstflint IMO no ?

wmousa replied:
Is there a way to do vf trust using mstflint ?! @adrianchiris

wmousa replied:
for now, just using mlxreg

adrianchiris commented on switchdev_utils/config-service/README.md:
```diff
@@ -0,0 +1,55 @@
+
+## Prerequisites
+The below is assumed to be installed/configured on the host
+- NICs Firmware up to date  (as bundled with MLNX_OFED 5.2)
```
drop the 5.2 here as we say a line below 5.2 or above

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/README.md:
```diff
@@ -0,0 +1,55 @@
+The below is assumed to be installed/configured on the host
+- NICs Firmware up to date  (as bundled with MLNX_OFED 5.2)
+- (MLNX_OFED 5.2 or above ) or (mft 4.16 or above)
+- NICs NVCONFIG configures SR-IOV to support VFs per PF
```
can you specify the commnad ?

e.g
`mstconfig -d <PCI device> set SRIOV_EN=True NUM_OF_VFS=<max_num_of_vfs>`

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/README.md:
```diff
@@ -0,0 +1,55 @@
+- ovsBridges:  
+List of ovs bridges which need to be created and configured  
+    - name: The name of ovs bridge  
+    - ports: List of ovs bridge ports (sriov pfs or linux bonds)  
```
nit: (sriov pfs or linux bonds) -> (a subset of sriov pfs or linux bonds)

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/README.md:
```diff
@@ -0,0 +1,55 @@
+    - name: The name of ovs bridge  
+    - ports: List of ovs bridge ports (sriov pfs or linux bonds)  
+
+## Setup  
```
as discussed, lets have `install.sh` and `uninstall.sh`  (its kina weird to have uninstall script without install script )

adrianchiris replied:
we should imo not put it directly in `/etc/` but in `/etc/switchdev-config/` WDYT ?

`install.sh` should get as input the config.yaml and will
1. copy switchdev_network_config
2. copy config file to /etc/switchdev-config/

that way we also take ownership of this config file
then `uninstall.sh` will delete it as well.

wmousa replied:
install  should just copy if the user provide a config and override the old one , and in uninstall don't remove the config just remove the script in order to preserve user config

adrianchiris commented on switchdev_utils/config-service/README.md:
````diff
@@ -0,0 +1,55 @@
+```$ vi /etc/config.yaml```  
+- Run the script  
+```$ /etc/switchdev_network_config```  
+After running the script you will have sriov vfs created, moved to switchdev, vfs representors created(if it's configure), linux bonds and   openvswitch bridges are up and configured.  
````
The script will perform the following:
1.
2.
3.
4.
...

wmousa replied:
it's already mentioned above, removing it

adrianchiris commented on switchdev_utils/config-service/README.md:
````diff
@@ -0,0 +1,55 @@
+
+## Uninstall
+Uninstall will remove "/etc/switchdev_network_config" script and also will remove sriov_config and sriov_bind services  
+```$ bash -x ./uninstall.sh```
````
why bash -x ? thats just for debug purposes

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,494 @@
+            logger.error("Error parsing config file: %s" % CONFIG_FILE)
+            exit(1)
+    else:
+        logger.error("Configuration file is not exist.. Exiting")
```
is not -> does not

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,494 @@
+    return add_udev_rule(udev_data_line, UDEV_RULE_FILE)
+
+
+def run_command(*cmd):
```
can you move this to the top of the file, as the whole script uses it, i think it should be first for code clarity

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,494 @@
+        ret = p.returncode
+        if ret != 0:
+            logger.error("Falied to execute %s \n%s" % (' '.join(cmd), err))
+            raise
```
what are you raising here ? you should raise a RuntimeError with a meaningful string imo

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,494 @@
+
+    for file in listOfPfFiles:
+        if file.startswith("virtfn"):
+
```
un-needed newline

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/uninstall.sh:
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+# Removing sriov_config service
+systemctl dinable sriov_config
```
dinable -> disable

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,494 @@
+bridges = []
+
+logging.basicConfig(filename=LOG_FILE,
+                    level=logging.DEBUG)
```
default logging level should be INFO

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,494 @@
+    if os.path.exists(CONFIG_FILE):
+        config_data = get_file_data(CONFIG_FILE)
+        try:
+            config_array = yaml.safe_load(config_data)
```
its a dict not an array right ? as i see in main you do `.get`

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,494 @@
+
+
+def main(argv):
+    config_array = get_config()
```
config_array is not an array just call it `config`

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,494 @@
+    config_array = get_config()
+    sriov_pfs = config_array.get("pfs")
+    bonds = config_array.get("linuxBonds")
+    bridges = config_array.get("ovsBridges")
```
you are now obscuring the global variables for sriov_pfs, bonds, bridges.

you are not using the globally scoped one, so you should remove any global variable that is not used.

wmousa replied:
done

moshe010 commented on switchdev_utils/config-service/uninstall.sh:
```diff
@@ -0,0 +1,12 @@
+rm -f "/etc/systemd/system/sriov_bind.service"
+
+# Removing switchdev_network_config
+rm -f "etc/switchdev_network_config"
```
missing "/"

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/uninstall.sh:
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
```
some basic indication to the user using echo to tell the user we are uninstalling and that uninstall it done would be helpful

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/README.md:
```diff
@@ -0,0 +1,64 @@
+## Install  
+Prepare you own config.yaml file with all the configuration you need and pass it to installation script or you can copy it directly to /etc/switchdev-config/config.yaml before execution  
+Install will do the following
+- create switchdev-config directory
```
create `/etc/switchdev-config` directory

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/README.md:
```diff
@@ -0,0 +1,64 @@
+Prepare you own config.yaml file with all the configuration you need and pass it to installation script or you can copy it directly to /etc/switchdev-config/config.yaml before execution  
+Install will do the following
+- create switchdev-config directory
+- copy switchdev_network_config to /etc/switchdev-config directory  
```
optionally copy configuration file to `/etc/switchdev-config` directory

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/README.md:
````diff
@@ -0,0 +1,64 @@
+Run the script  
+```$ bash ./install.sh [config.yaml]```  
+
+## Execution
````
thats first time execution or on configuration change ? if so you should say this

wmousa replied:
on configuration change, done

adrianchiris commented on switchdev_utils/config-service/README.md:
```diff
@@ -0,0 +1,64 @@
+- create switchdev-config directory
+- copy switchdev_network_config to /etc/switchdev-config directory  
+
+Run the script  
```
nit: just - Run:

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/README.md:
````diff
@@ -0,0 +1,64 @@
+- disable and remove sriov_config service  
+- disable and remove sriov_bind service  
+
+```$ bash ./uninstall.sh```
````
Run:
`<uninstall command>`

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/sample_config.yaml:
```diff
@@ -0,0 +1,30 @@
+  - name: ens1f1
+    numofvf: 8
+    switchdev: False
+
```
nit: remove newline

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,548 @@
+        out, err = p.communicate()
+        ret = p.returncode
+        if ret != 0:
+            err_msg = "Falied to execute %s \n%s" % (' '.join(cmd), err)
```
not sure how nice the `\n` will look in log, why not just :  `Failed to execute: %s. %s`

wmousa replied:
new line to separate the command and the error

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,548 @@
+        logger.debug(out)
+        return out
+    except Exception as e:
+        logger.error("Failed to execute %s\n%s" % (' '.join(cmd), str(e)))
```
same comment as above with the `\n`

wmousa replied:
new line to separate the command and the error

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,548 @@
+    bonds = config.get("linuxBonds")
+    bridges = config.get("ovsBridges")
+    hw_offload_enabled = config.get("hwOffloadEnabled")
+    validate_config(sriov_pfs, bonds, bridges)
```
validate_config should do what it says validate the configuration file, IMO it should run after L509 and get `config` as input.

one of its validations should be the config version, and fail if `configVersion.major` > 1
also emit warning if `configVersion.minor` > 0 as it may indicate new optional functionality this script does not know how to handle.

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,548 @@
+            raise ValueError(err)
+
+
+def main(argv):
```
so no argparse ? :)

we should have imo at least a version flag so we know what is installed.

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,548 @@
+            logger.error(err)
+            raise ValueError(err)
+        try:
+            if int(sriov_pf.get("numOfVfs")):
```
-17 is also an integer but its value is invalid in this context

wmousa replied:
The validation is meant just if the user forgot to configure something by mistake

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,548 @@
+    sriov_pfs_list = []
+    bonds_list = []
+    for sriov_pf in sriov_pfs:
+        if sriov_pf.get("name"):
```
name must be a string

wmousa replied:
The validation is meant just if the user forgot to configure something by mistake

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,548 @@
+
+
+def validate_config(sriov_pfs, bonds, bridges):
+    sriov_pfs_list = []
```
i realize this method is not the heart and soul of this PR, but we should fortify it a bit more.
it will not be 100% bullet proof but will be "good enough".

if we end up exenting the config we should consider defining a schema for it

wmousa replied:
I agree but just for now small validation for critical stuff

wmousa replied:
The validation is meant just if the user forgot to configure something by mistake

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,548 @@
+            logger.error("numOfVfs for %s must be provided" %
+                         sriov_pf.get("name"))
+    for bond in bonds:
+        if bond.get("name"):
```
name must be a string

adrianchiris replied:
as well as the rest of the attributes below.

wmousa replied:
The validation is meant just if the user forgot to configure something by mistake

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,548 @@
+            logger.error(err)
+            raise ValueError(err)
+        if (bridge.get("ports") and
+                (set(bridge.get("ports")).issubset(set(sriov_pfs_list +
```
technically if a PF is a member of a bond, it must not be a member of a bridge right ?

wmousa replied:
yes but here I'm not going to validate it all just to make sure that the user didn't write something by mistake, change letter, name..etc

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,548 @@
+import subprocess
+import sys
+import time
+import yaml
```
`yaml` needs to be installed or marked as a pre-req in README

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/README.md:
```diff
@@ -0,0 +1,76 @@
+    - ports: List of ovs bridge ports (a subset of sriov pfs or linux bonds)  
+
+## Install  
+Prepare you own config.yaml file with all the configuration you need and pass it to installation script or you can copy it directly to /etc/switchdev-config/config.yaml before execution  
```
your own

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/README.md:
```diff
@@ -0,0 +1,76 @@
+## Install  
+Prepare you own config.yaml file with all the configuration you need and pass it to installation script or you can copy it directly to /etc/switchdev-config/config.yaml before execution  
+Install will do the following
+- create /etc/switchdev-config directory  
```
for every file can you add single quotes like `this` ?

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/README.md:
````diff
@@ -0,0 +1,76 @@
+Run:  
+```$ bash /etc/switchdev-config/uninstall.sh```
+**Note:**
+- Current configuration for sriov pfs will stay exist until you do reboot
````
SR-IOV PF configuration will reset after reboot

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/README.md:
````diff
@@ -0,0 +1,76 @@
+```$ bash /etc/switchdev-config/uninstall.sh```
+**Note:**
+- Current configuration for sriov pfs will stay exist until you do reboot
+- network configuration (bonds and bridges) will stay after reboot unless you clean them manually:
````
each should start with a capital letter

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,678 @@
+
+
+def parse_opts(argv):
+
```
nit: no need for newline

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,678 @@
+
+def main(argv=sys.argv):
+    opts = parse_opts(argv)
+
```
nit: remove newline

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,678 @@
+    parser.add_argument(
+        '-v', '--version',
+        action='version',
+        version="%d.%d" % (VERSION_MAJOR, VERSION_MINOR))
```
I would really prefer a Major.Minor.SubMinor version

so that:
Major: marks a breaking change
Minor: is a release with new features
SubMinor is generally bugfixes to current release with no new features added

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,678 @@
+        '-v', '--version',
+        action='version',
+        version="%d.%d" % (VERSION_MAJOR, VERSION_MINOR))
+
```
need to test `opts.sriov_bind == true and opts.sriov_config == true` and fail as these are mutually exclusive as i see from the code.

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,678 @@
+    # validate_config is just a quick validation indicates if the user has
+    # typo or missing mandatory prams
+    # TODO//
+    # Define a schema for config.yaml file to do full config validation
```
some validations cannot be checked in schema and will need to be validated in code.

what i wanted to add here is essentialy a TODO with missing validations.

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,678 @@
+    bonds_list = []
+    version = config.get("configVersion")
+    version_major, version_minor = parse_version(version)
+    if version_major > VERSION_MAJOR:
```
There is no 1:1 relation between the app version and the config version.

what if you need to push as minor fix to this code but the config version does not change ?

adrianchiris replied:
There should be separate version for the script/app and the config version this script/app supports

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,678 @@
+    version = config.get("configVersion")
+    version_major, version_minor = parse_version(version)
+    if version_major > VERSION_MAJOR:
+        err = "Provided Major version is upper than current one. Exiting.."
```
Provided configuration version is not supported.

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,678 @@
+        logger.error(err)
+        raise(RuntimeError(err))
+    if version_minor > VERSION_MINOR:
+        logger.debug("Provided Minor version is upper than current one, new"
```
"Provided configuration version newer, new functionality may not be handled by this script"

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,678 @@
+    version_major = 0
+    version_minor = 0
+    try:
+        version_float = float(version)
```
isnt the version a string (you can convert to string if needed)? you could simply split the string with a `.` and extract the fields

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,678 @@
+    with open(REP_LINK_NAME_FILE, "w") as f:
+        f.write(REP_LINK_NAME_DATA)
+    # Make the _REP_LINK_NAME_FILE executable
+    os.chmod(REP_LINK_NAME_FILE, 0o755)
```
0o755 ?

shouldnt it be 755 ? or 0755 ?

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,678 @@
+                                                 pf_fun_num,
+                                                 REP_LINK_NAME_FILE,
+                                                 sriov_pf)
+    create_rep_link_name_script()
```
L271 need to happen once, however this function is called for every sriov PF

wmousa replied:
done

adrianchiris commented on switchdev_utils/config-service/switchdev_network_config:
```diff
@@ -0,0 +1,678 @@
+
+
+def add_udev_rule_to_unmanage_vf_representors_by_nm():
+    udev_data_line = 'SUBSYSTEM=="net", ACTION=="add", ATTR{phys_switch_id}'\
```
This will unmanage all vf representors, regardless if their PF was mentioned in the config file.

is that how os-net-config work ?

wmousa replied:
yes

wmousa committed:
This repo describes the sriov and network configuration that needs
to occur on any node that need to configure sriov_pfs with switchdev
capability and vf-lag with bonding

wmousa committed:
Switchdev Network Configure

This repo describes the sriov and network configuration that needs
to occur on any node that need to configure sriov_pfs with switchdev
capability and vf-lag with bonding

wmousa committed:
Switchdev Network Configure

This repo describes the sriov and network configuration that needs
to occur on any node that need to configure sriov_pfs with switchdev
capability and vf-lag with bonding

wmousa commented:
@adrianchiris, It's ready now
